clar2wasm/src/wasm_utils.rs

```diff
@@ -1285,17 +1286,22 @@ pub fn call_function<'a>(
     // Call the function
     func.call(&mut store, &wasm_args, &mut results)
         .map_err(|e| {
-            // TODO: If the root cause is a clarity error, we should be able to return that,
-            //       but it is not cloneable, so we can't return it directly.
-            //       If the root cause is a trap from our Wasm code, then we need to translate
-            //       it into a Clarity error.
-            //       See issue stacks-network/clarity-wasm#104
-            // if let Some(vm_error) = e.root_cause().downcast_ref::<crate::vm::errors::Error>() {
-            //     vm_error.clone()
-            // } else {
-            //     Error::Wasm(WasmError::Runtime(e))
-            // }
-            Error::Wasm(WasmError::Runtime(e))
+            if let Some(vm_error) = e.root_cause().downcast_ref::<Error>() {
+                // This unsafe operation returns the value of a location pointed by `*mut T`.
+                // The purpose of this code is to take the ownership of the `vm_error` value.
+                // The replaced `T` value is deallocated after the operation. Therefore, the chosen `T`
+                // is a dummy value, solely to satisfy the signature of the replace function
+                // and not cause harm when it is deallocated.
+                // Specifically, Error::Wasm(WasmError::ModuleNotFound) was selected as the placeholder value.
```

Comment:
- I don't know why this doesn't trigger [**undocumented unsafe blocks**](https://rust-lang.github.io/rust-clippy/master/index.html#/undocumented_unsafe_blocks), but this comment should start with `SAFETY:`.
- There should be an explanation to why converting a non-mutable reference to a mutable pointer doesn't cause any issue here. (e.g. is there any possibility that this error is borrowed by anything else at the same time?)
- I think the comment should mention that the reason we need this hack is because `Error` is not `Clone` and we need `replace` to avoid a potential double free.

Reply:
I think we need a solution to this that does not require any unsafe code. Unsafe code is not likely to be acceptable as part of stacks-core.

Reply:
Hey @obycode ! Thanks for chiming in.

In my updated comment, I addressed the unsafe operation. It's essentially a workaround to take ownership of the `Error` reference, but it's harmless because the newly allocated value can be safely deallocated.

But since we cannot have unsafe blocks, it seems like other parts of the codebase relying on unsafe operations might also require changes.

Anyway, I am open to discuss and implement new solutions. Tks!

Reply:
It's not that we can't have unsafe blocks... we can if we really need to and we can be sure that it is safe. I think the only use currently (not in tests) is this:

```rust
    let module = context
        .contract_context()
        .with_wasm_module(|wasm_module| unsafe {
            Module::deserialize(&engine, wasm_module)
                .map_err(|e| Error::Wasm(WasmError::UnableToLoadModule(e)))
        })?;
```

This one though should have a comment added clarifying its safety.

Reply:
(also, I think this function calling mechanism ultimately needs to be replaced with something better, but that is a totally separate issue)